src/prediction_market/views/pred_market_votes_view.py

```diff
@@ -0,0 +1,181 @@
+                return Response(
+                    {"message": "prediction_market_id or paper_id is required"},
+                    status=400,
+                )
+
+            # get prediction market, or create one if it doesn't exist
+            prediction_market = get_or_create_prediction_market(paper_id)
+            pred_mkt_id = prediction_market.id
+
+        try:
+            prediction_market = PredictionMarket.objects.get(id=pred_mkt_id)
+        except PredictionMarket.DoesNotExist:
+            return Response({"message": "Prediction market does not exist"}, status=400)
+
+        # check if user has voted before
+        prev_vote = PredictionMarketVote.objects.filter(
+            created_by=user, prediction_market=prediction_market
+        ).first()
+
+        if prev_vote is not None:
+            # update vote
+            prev_vote.vote = vote
+            prev_vote.bet_amount = bet_amount
+            prev_vote.save()
+            prediction_market_vote = prev_vote
+        else:
+            # create vote
+            prediction_market_vote = PredictionMarketVote.objects.create(
+                created_by=user,
+                prediction_market=prediction_market,
+                vote=vote,
+                bet_amount=bet_amount,
+            )
+
+            # track as contribution if it's a new vote
+            create_contribution.apply_async(
+                (
+                    Contribution.REPLICATION_VOTE,
+                    {
+                        "app_label": "prediction_market",
+                        "model": "predictionmarketvote",
+                    },
+                    request.user.id,
+                    prediction_market.unified_document.id,
+                    prediction_market_vote.id,
+                ),
+                priority=1,
+                countdown=10,
+            )
+
+        context = self._get_retrieve_context()
+        data = DynamicPredictionMarketVoteSerializer(
+            prediction_market_vote, context=context
+        ).data
+        return Response(data, status=200)
+
+    def _get_retrieve_context(self):
+        context = self.get_serializer_context()
+        context = {
+            **context,
+            "rhc_dcs_get_created_by": {
+                "_include_fields": (
+                    "id",
+                    "author_profile",
+                    "first_name",
+                    "last_name",
+                    "editor_of",
+                )
+            },
+            "usr_dus_get_author_profile": {
+                "_include_fields": (
+                    "id",
+                    "first_name",
+                    "last_name",
+                    "created_date",
+                    "updated_date",
+                    "profile_image",
+                )
+            },
+        }
+        return context
+
+    def list(self, request, *args, **kwargs):
+        pred_mkt_id = request.query_params.get("prediction_market_id")
```

Comment:
In a ModelViewSet the list functionality is built in. 

If we just want to query by custom query parameters, you can look at django-filters (we use it through the rest of our projects too). 

Makes the code a bit easier to understand by fitting into that format